### ISSUE TITLE
overzealous out of order decl compiler error

### ISSUE THREAD

**Issue by gewang** [ccrma/chuck#376]
encountering compiler error
```
iss-out-of-order.ck:6:13: error: variable/member 'a' is used before declaration
[6]         <<< a >>>;
                ^
iss-out-of-order.ck: ...in function 'void Foo.go()'
```
in the following program:
```
class Foo
{
    fun void go()
    {
        // this shouldn't produce an error, but does
        <<< a >>>;
    }
}

// decl and init
5 => int a;
```
possible problem: overzealous checking of decl order
possible remedy: should handle global usage and func/classdef usage of `a` differently

**Comment by gewang** [comment]
another example
```
public class Foo
{
    BarFactory.make() @=> Bar bar;
}

class Bar
{
    "i am bar" => string name;
}

class BarChild extends Bar
{
    "i extend bar" => name;
}

class BarFactory
{
    fun static BarChild make()
    {
        return new BarChild;
    }
}
```

**Comment by gewang** [comment]
addressed in dbcec841135bd9f24c7357ccfb1f50c542443c4b and b6206584314c1b8237264eddb82c85858fcf0698, included in 1.5.2.0
